src/Compilers/Core/Portable/Collections/Rope.cs

```diff
@@ -0,0 +1,151 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// A representation of a string of characters that requires O(1) extra space to concatenate two ropes.
+    /// </summary>
+    internal abstract class Rope
+    {
+        public static readonly Rope Empty = ForString("");
+        public abstract override string ToString();
+        public abstract int Length { get; }
+        protected abstract IEnumerable<char> GetChars();
+
+        /// <summary>
+        /// A rope can wrap a simple string.
+        /// </summary>
+        public static Rope ForString(string s)
+        {
+            if (s == null)
+                throw new ArgumentNullException(nameof(s));
+
+            return new StringRope(s);
+        }
+
+        /// <summary>
+        /// A rope can be formed from the concatenation of two ropes.
+        /// </summary>
+        public static Rope Concat(Rope r1, Rope r2)
+        {
+            if (r1 == null)
+                throw new ArgumentNullException(nameof(r1));
+
+            if (r2 == null)
+                throw new ArgumentNullException(nameof(r2));
+
+            return
+                r1.Length == 0 ? r2 :
+                r2.Length == 0 ? r1 :
+                checked(r1.Length + r2.Length < 32) ? ForString(r1.ToString() + r2.ToString()) :
+                new ConcatRope(r1, r2);
+        }
+
+        /// <summary>
+        /// Two ropes are "the same" if they represent the same sequence of characters.
+        /// </summary>
+        public override bool Equals(object obj)
+        {
+            if (!(obj is Rope other) || Length != other.Length)
+                return false;
+            if (Length == 0)
+                return true;
+            var chars0 = GetChars().GetEnumerator();
+            var chars1 = other.GetChars().GetEnumerator();
```

Comment:
Implementing `GetChars` in terms of `Span<char>` would avoid allocations in this method and let you use `Span<T>.SequenceEquals` which is very efficient on CoreClr